File: src_mcp/runners/Reviewers/RemoteAsk.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CoaiMcp.Runners.Reviewers;

/// <summary>What a poll of a Team server review came back as.</summary>
public enum RemoteState
{
    Queued,
    Running,
    Done,
    Failed,

    /// <summary>
    /// A status this client has no branch for.
    /// </summary>
    /// <remarks>
    /// It is NOT folded into <see cref="Queued"/>, which is what it used to be. A newer server adding
    /// a terminal status would then have been read as "still waiting", and the shim would have polled
    /// a finished review until its own deadline and reported it as too slow — a wrong sentence about
    /// the wrong thing. Raised on the code round.
    /// </remarks>
    Unknown,
}

/// <summary>One poll's answer, parsed.</summary>
/// <param name="Position">Where it sits in its vendor's queue; 0 once it is running.</param>
/// <param name="Answer">The vendor's RAW text. This shim does not parse it — the round does.</param>
public sealed record RemotePoll(
    RemoteState State,
    string RawStatus,
    int Position,
    string Answer,
    string Failure,
    string Reason,
    long TokensIn,
    long TokensOut);

/// <summary>
/// The pure half of <c>--ask-remote</c>: what to send, how to read what comes back, and every
/// sentence a person will see.
/// </summary>
/// <remarks>
/// The twin of <see cref="LocalAsk"/>, and pure for the same reason: every exit path can be tested
/// without a server, so the wording is checked rather than hoped for. A sentence that blames the
/// wrong thing sends somebody to fix something that was never broken.
/// </remarks>
public static class RemoteAsk
{
    /// <summary>Exit codes. The same vocabulary <c>--ask-local</c> uses.</summary>
    public const int Ok = 0;
    public const int BadUsage = 64;
    public const int MissingSchema = 65;
    public const int Unreachable = 69;
    public const int VendorFailed = 70;
    public const int TooOld = 76;
    public const int NotSignedIn = 77;

    /// <summary>
    /// The review did not finish in time, and has been cancelled.
    /// </summary>
    /// <remarks>
    /// Its own code rather than <see cref="Unreachable"/>, which is what it used to return. The
    /// sentence said "did not finish within 90s" while the exit code said the server could not be
    /// reached — so anybody reading the code rather than the text went to check their network for a
    /// problem whose cure is a longer timeout or more accounts. Raised on the code round.
    /// </remarks>
    public const int TooSlow = 78;

    /// <summary>How long to leave between polls while a review is queued or running.</summary>
    /// <remarks>
    /// The server long-polls up to 25 seconds and returns the moment the state changes, so this is
    /// how long to wait after a poll that came back unchanged — not how often the server is asked.
    /// One reviewer therefore makes about two requests a minute, which is why polling never reaches
    /// the server's rate limiter.
    /// </remarks>
    public static readonly TimeSpan PollGap = TimeSpan.FromSeconds(1);

    /// <summary>The header every request carries, and what this client speaks.</summary>
    /// <remarks>
    /// The server judges this BEFORE it looks at the token, so a client too old to be served is told
    /// to update rather than handed a 401 about a token that was never the problem. Declared here
    /// rather than imported from the server: these are two binaries that ship separately, and a
    /// shared constant would mean the client could only be as old as the server it was built beside
    /// — which is the exact thing the header exists to make unnecessary.
    /// </remarks>
    public const string ContractHeader = "X-Coai-Contract";

    /// <inheritdoc cref="ContractHeader"/>
    public const int ContractVersion = 1;

    /// <summary>How long each long poll asks the server to hold the connection.</summary>
    public const int LongPollSeconds = 25;

    /// <summary>The body of <c>POST /api/reviews</c>.</summary>
    public static string RequestBody(string vendor, string model, string role, string prompt, int timeoutSeconds) =>
        JsonSerializer.Serialize(
            new RemoteRequest(vendor, model, prompt, role, timeoutSeconds),
            RemoteJsonContext.Default.RemoteRequest);

    /// <summary>The id in a 202, or empty when the answer was not one.</summary>
    public static string AcceptedId(string body)
    {
        try
        {
            return JsonSerializer.Deserialize(body, RemoteJsonContext.Default.RemoteAccepted)?.Id ?? string.Empty;
        }
        catch (JsonException)
        {
            return string.Empty;
        }
    }

    /// <summary>One poll's answer, or null when the body was not one.</summary>
    public static RemotePoll? ReadPoll(string body)
    {
        try
        {
            var status = JsonSerializer.Deserialize(body, RemoteJsonContext.Default.RemoteStatus);

            return status is null
                ? null
                : new RemotePoll(
                    StateOf(status.Status),
                    status.Status ?? string.Empty,
                    status.Position,
                    status.Answer ?? string.Empty,
                    status.Failure ?? string.Empty,
                    status.Reason ?? string.Empty,
                    status.TokensIn,
                    status.TokensOut);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    /// <summary>The usage line the runtime reads back off stdout.</summary>
    public static string UsageLine(long tokensIn, long tokensOut) =>
        JsonSerializer.Serialize(new RemoteUsage(tokensIn, tokensOut, null), RemoteJsonContext.Default.RemoteUsage);

    /// <summary>Not signed in at all — there is no token file for this server.</summary>
    public static string NotSignedInMessage(string serverUrl) =>
        $"not signed in to the Team server at {serverUrl} — sign in from the panel's Team servers section";

    /// <summary>Signed in once, but the server will not take this token any more.</summary>
    public static string RejectedMessage(string serverUrl) =>
        $"the Team server at {serverUrl} rejected this token — sign in again from the panel's Team servers section";

    /// <summary>Authenticated, and not welcome.</summary>
    /// <remarks>
    /// A different sentence from the one above on purpose: signing in again cannot fix being outside
    /// the company's domain, and telling somebody to try would waste their afternoon.
    /// </remarks>
    public static string ForbiddenMessage(string serverUrl) =>
        $"the Team server at {serverUrl} refused this account — it is outside the allowed company domain";

    public static string TooOldMessage(string serverUrl, string serverSaid) =>
        $"coai-mcp is older than the Team server at {serverUrl} requires — update it from the panel's "
        + $"Server section. The server said: {Trim(serverSaid, 200)}";

    public static string UnreachableMessage(string serverUrl, string detail) =>
        $"the Team server at {serverUrl} could not be reached: {Trim(detail, 200)}";

    /// <summary>
    /// The server answered something this shim has no branch for.
    /// </summary>
    /// <remarks>
    /// A 500, a 404 from a proxy that swallowed the route, an HTML error page from something in
    /// front of it. Left unhandled these become a crash or a hang; named, they become one sentence
    /// with the status in it, which is the difference between "the server is broken" and "something
    /// in the middle is". (codex, plan round.)
    /// </remarks>
    public static string UnexpectedMessage(string serverUrl, int status, string body) =>
        $"the Team server at {serverUrl} answered {status}, which this client has no handling for. "
        + $"It said: {Trim(body, 200)}";

    /// <summary>A body that should have been JSON and was not.</summary>
    public static string UnreadableMessage(string serverUrl, string body) =>
        $"the Team server at {serverUrl} answered something this client could not read as a review. "
        + $"It said: {Trim(body, 200)}";

    /// <summary>The shim gave up. The job is cancelled before this is printed.</summary>
    /// <remarks>
    /// It names the position AND what happens next, because "position 50" alone tells somebody
    /// nothing they can act on — they cannot tell five minutes from five hours, and the usual
    /// response to not knowing is to retry, which lengthens the queue they are complaining about.
    /// (local, plan round.)
    /// </remarks>
    public static string TooSlowMessage(string serverUrl, TimeSpan waited, int position) =>
        $"the review did not finish within {waited.TotalSeconds:0}s on the Team server at {serverUrl}"
        + (position > 0 ? $" (it was {position} in the queue)" : " (it had started)")
        + ". It has been cancelled, so nothing is still running. Raise COAI_REVIEWER_TIMEOUT_MINUTES, "
        + "or ask the operator for more accounts on that vendor.";

    /// <summary>A status from a server this client is too old to understand.</summary>
    public static string UnknownStateMessage(string serverUrl, string status) =>
        $"the Team server at {serverUrl} reported a review status this client does not know "
        + $"('{Trim(status, 60)}') — update coai-mcp from the panel's Server section";

    /// <summary>The vendor itself failed, in its own words.</summary>
    public static string FailedMessage(string vendor, string failure, string reason) =>
        $"the Team server's {vendor} reviewer failed ({(failure.Length > 0 ? failure : "no reason given")})"
        + (reason.Length > 0 ? $": {Trim(reason, 400)}" : string.Empty);

    /// <summary>Still waiting for a free account, and how many are ahead.</summary>
    public static string QueuedMessage(string serverUrl, int position) =>
        $"queued on the Team server at {serverUrl}, {position} ahead of it";

    /// <summary>On an account now, running.</summary>
    public static string RunningMessage(string serverUrl) =>
        $"running on the Team server at {serverUrl}";

    /// <summary>
    /// Whether a line the shim wrote is PROGRESS rather than a verdict.
    /// </summary>
    /// <remarks>
    /// <para>The shim writes both to one stderr, and whoever reads that stream to explain a failure
    /// has to be able to tell them apart. Measured on a real gate, 2026-09-07: a Team-server
    /// reviewer failed and the round reported
    /// <c>exit 70: [coai-mcp] claude: running on the Team server at …</c> — a reviewer described as
    /// RUNNING in the sentence announcing that it had stopped, because the reason-picker took the
    /// first line that was not obviously noise and a progress note is the first line there is.</para>
    /// <para><b>Asking "is this progress" rather than "does this announce a failure" is the whole
    /// point.</b> The first attempt taught the picker the word <c>failed</c>, and three reviewers
    /// independently broke it: <c>0 failed, 3 passed</c> and <c>3 tests failed</c> are tallies that
    /// announce nothing, while <c>Step 3 failed</c> and <c>Job 12 failed</c> are real verdicts with
    /// a digit in front — no rule over that word can have it both ways. These two sentences, by
    /// contrast, are ours: we author them, there are exactly two, and neither can be confused with
    /// anything a vendor CLI prints.</para>
    /// <para>Matched on the STEM rather than the whole sentence, because the server URL and the
    /// queue position are interpolated into them.</para>
    /// </remarks>
    public static bool IsProgress(string line) =>
        line.Contains("queued on the Team server", StringComparison.OrdinalIgnoreCase) ||
        line.Contains("running on the Team server", StringComparison.OrdinalIgnoreCase);

    /// <summary>Which exit an HTTP status means, for a status this shim knows.</summary>
    /// <remarks>
    /// Returns null when the status is not one of the handled ones, so the caller can produce the
    /// "no handling for" sentence rather than guessing at an exit code.
    /// </remarks>
    public static int? ExitForStatus(int status) => status switch
    {
        401 => NotSignedIn,
        403 => NotSignedIn,
        426 => TooOld,
        >= 200 and < 300 => Ok,
        _ => null,
    };

    private static RemoteState StateOf(string? status) => status?.ToLowerInvariant() switch
    {
        "done" => RemoteState.Done,
        "failed" => RemoteState.Failed,
        "running" => RemoteState.Running,
        "queued" => RemoteState.Queued,
        _ => RemoteState.Unknown,
    };

    private static string Trim(string text, int max)
    {
        var single = (text ?? string.Empty).Replace('\n', ' ').Replace('\r', ' ').Trim();

        return single.Length <= max ? single : single[..max] + "…";
    }
}

internal sealed record RemoteRequest(string Vendor, string Model, string Prompt, string Role, int TimeoutSeconds);

internal sealed record RemoteAccepted(string? Id, int Position);

internal sealed record RemoteStatus(
    string? Id,
    string? Status,
    int Position,
    string? Answer,
    double Seconds,
    long TokensIn,
    long TokensOut,
    string? Failure,
    string? Reason);

internal sealed record RemoteUsage(long TokensIn, long TokensOut, double? CostUsd);

internal sealed record RemoteError(string? Error);

/// <summary>Every shape this shim puts on or takes off the wire, source-generated.</summary>
/// <remarks>
/// The server serialises camelCase and reflection is off in the AOT build of both binaries, so the
/// naming policy is declared here rather than left to a default that differs between them.
/// </remarks>
[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase, WriteIndented = false)]
[JsonSerializable(typeof(RemoteRequest))]
[JsonSerializable(typeof(RemoteAccepted))]
[JsonSerializable(typeof(RemoteStatus))]
[JsonSerializable(typeof(RemoteUsage))]
[JsonSerializable(typeof(RemoteError))]
internal sealed partial class RemoteJsonContext : JsonSerializerContext;

File: src_vs_code/src/remoteAsk.ts
/**
 * What a Team server is asked for a chat turn, and how its answers are read.
 *
 * <p>The protocol is not new: `coai-mcp --ask-remote` has spoken it since the Team server existed,
 * and the shapes here are that implementation's — `src_mcp/runners/Reviewers/RemoteAsk.cs`. Writing
 * a second dialect of one wire is how two halves become green about opposite things, which this
 * repository has paid for once already; so this file follows the C# one deliberately, and the names
 * are its names.</p>
 *
 * <p><b>A chat turn IS a review job to the server.</b> One endpoint, one queue, one accounting. What
 * differs is that it carries NO role — see {@link CHAT_ROLE}, which is a measurement rather than a
 * choice — so that a person reading their spending can tell a conversation from a round, because
 * every review carries one and nothing else does.</p>
 *
 * <p>Pure: every decision below is a test rather than a claim, and the socket is somebody else's job.</p>
 */

/**
 * The role a chat turn carries: NONE, and that is the measurement rather than an omission.
 *
 * <p>The first version sent `Chat`, which reads well and the server refuses outright: *"'Chat' is not
 * a review role. Allowed: PlanCritique, Conventions, Architecture, SecurityReliability,
 * UxDxPerformance"*. Measured against the live server on 2026-09-09, in 0.2 seconds, before a line
 * of the wiring was trusted.</p>
 *
 * <p>And the server is right. A review role is not a label — it carries a shipped prompt, a
 * threshold and a round budget, and a test in `src_mcp` walks every value of that enum asserting
 * each one asks for an honest empty findings list. A chat has none of those things, so adding it
 * there would make a reviewer that cannot review.</p>
 *
 * <p>What the server DOES accept is a job with no role at all — the validation skips the enum check
 * for an empty one — and that is exactly the separation the owner asked for: a usage row with no
 * role is a conversation, because every review carries one. Measured the same afternoon: empty role,
 * a model the catalog allows, and the answer came back.</p>
 *
 * <p>A `kind` field would say it better than an absence does, and it is a server change with its own
 * release. Recorded as the follow-up rather than smuggled into the role enum.</p>
 */
export const CHAT_ROLE = '';

/**
 * How many turns a REMOTE conversation may hold.
 *
 * <p>The owner's ruling, asked and answered on 2026-09-08: *"Стоит ли вообще запрещать длинные
 * разговоры на серверных моделях макс 3 хода"* — yes. A remote model keeps no conversation, so every
 * turn re-sends the whole transcript, and the cost of turn N is the cost of everything before it. An
 * unbounded conversation is therefore a bill that grows quadratically and, eventually, a request the
 * server refuses for its size. Three turns, said in the page before the fourth is typed, and the
 * local model offered instead — because a local one HAS a memory and no such limit.</p>
 */
export const REMOTE_TURNS = 3;

/** What one poll of a queued review means. */
export type RemoteStep =
  /** Still queued or running. `position` is where it sits, or 0 when the server did not say. */
  | { readonly kind: 'waiting'; readonly position: number }
  /**
   * The answer, and only the answer.
   *
   * <p>It carried the turn's token counts for a while and nothing ever read them: the server is the
   * system of record for usage — the job is a job in its ledger — and the panel's spending section
   * reads them from there. A number nobody renders is not free, because the day somebody renders it
   * a missing measurement will already have been rounded to `0`, which is the one thing a
   * measurement that could not be taken must never look like. (codex, the code round.)</p>
   */
  | { readonly kind: 'answer'; readonly text: string }
  | { readonly kind: 'failure'; readonly failure: string };

/**
 * The shape a review id may have before it is put into an authenticated URL.
 *
 * <p>The id comes from the server, and it goes into `api/reviews/<id>` on a request carrying this
 * machine's bearer token. A server that is compromised, or merely something answering in front of
 * one, could send `../../api/servers` or an id carrying query syntax and steer that token at a route
 * nobody chose. The same guard the catalog's vendor ids get in `isUsableVendorId`, and the thread
 * ids in the local adapters — three places, one rule: a stranger's identifier is checked against the
 * shape an identifier has before it becomes part of anything. (codex, the code round.)</p>
 */
const REVIEW_ID = /^[A-Za-z0-9][A-Za-z0-9._-]{0,99}$/;

/** The body a submit carries. Shaped by `RemoteRequest` on the server's side. */
export function requestBody(
  vendor: string,
  model: string,
  prompt: string,
  timeoutSeconds: number,
): Record<string, unknown> {
  return { vendor, model, prompt, role: CHAT_ROLE, timeoutSeconds };
}

/** The id a submit was accepted under, or empty when the answer was not one — or not an id. */
export function acceptedId(body: unknown): string {
  const row = body as { id?: unknown } | null;
  const said = typeof row === 'object' && row !== null && typeof row.id === 'string' ? row.id : '';

  return REVIEW_ID.test(said) ? said : '';
}

/**
 * What the server said about a review this poll.
 *
 * <p>The status strings are the server's, and an unknown one is treated as still running rather than
 * as a failure: a server that learns a new state must not turn every conversation into an error on
 * the day it ships. The deadline is what ends a turn that never resolves.</p>
 */
export function readStep(body: unknown): RemoteStep {
  const row = (body ?? {}) as Record<string, unknown>;
  const status = typeof row['status'] === 'string' ? row['status'].toLowerCase() : '';
  const answer = typeof row['answer'] === 'string' ? row['answer'] : '';
  const position = typeof row['position'] === 'number' ? row['position'] : 0;

  if (status === 'done' || status === 'completed' || status === 'succeeded') {
    return answer.trim().length > 0
      ? { kind: 'answer', text: answer.trim() }
      // A finished review with nothing in it is a failure, not an empty answer: a page showing
      // nothing looks like a model with nothing to say. The same rule as the local adapters'.
      : { kind: 'failure', failure: 'the server finished the turn without an answer' };
  }
  if (status === 'failed' || status === 'error' || status === 'cancelled' || status === 'canceled') {
    return { kind: 'failure', failure: `the server reported: ${saidWhy(row, status)}` };
  }
  if (status.length === 0 || RUNNING.includes(status)) {
    return { kind: 'waiting', position };
  }

  // The server has exactly four states — Queued, Running, Done, Failed — so a fifth value is a
  // server much newer than this build, or something in front of it answering for it. Three minutes
  // of "Thinking…" is the worst way to say either, and naming it is the shortest way to be fixed.
  // (gemini and local, one finding from two sides; the state list is measured, not assumed.)
  return { kind: 'failure', failure: `the server answered with a status this build does not know: ${status}` };
}

/** The states that mean "not finished yet". Measured from the server's own `JobStatus`. */
const RUNNING: readonly string[] = ['queued', 'running', 'pending', 'claimed', 'waiting'];

/** What the server said about a failure, in its own words when it gave any. */
function saidWhy(row: Record<string, unknown>, status: string): string {
  const failure = typeof row['failure'] === 'string' ? row['failure'] : '';
  const reason = typeof row['reason'] === 'string' ? row['reason'] : '';

  return failure.length > 0 ? failure : reason.length > 0 ? reason : status;
}

/**
 * How long to ask the server to hold one poll open.
 *
 * <p>Never longer than what is LEFT of the turn's deadline: asking for the full window when four
 * seconds remain means the answer arrives after this side has already given up, which reads as a
 * server that never answered. The C# shim learned this the same way.</p>
 *
 * <p><b>And never longer than this side will WAIT.</b> `ask` aborts every request at
 * `REQUEST_TIMEOUT_MS` — ten seconds — so a poll that asks the server to hold for twenty-five is a
 * request this side kills before the server answers: every poll failing with "it did not answer
 * within 10s", for a server behaving perfectly. Eight leaves room for the round trip. The C# shim
 * can ask for twenty-five because its own client has no such cap; this one cannot, and the two
 * constants had to be read together to see it. (codex, the plan round, from the other end.)</p>
 */
export function waitSecondsFor(remainingMs: number, longest = 8): number {
  const seconds = Math.floor(remainingMs / 1000);

  return Math.max(1, Math.min(longest, seconds));
}

/**
 * How long to wait after a refusal before asking again.
 *
 * <p>`429` is the shared-account ceiling doing its job — somebody else's round has the vendor — and
 * the answer is to wait, not to hammer. Doubling from a second, capped, so a queue that is briefly
 * full costs a second and one that is full for a minute costs a poll every ten.</p>
 */
export function backoffMs(attempt: number, capMs = 10_000): number {
  return Math.min(capMs, 1000 * 2 ** Math.max(0, attempt));
}

/**
 * Is this conversation full?
 *
 * <p>Counted in TURNS the person asked, not messages: a transcript holds a question and an answer
 * for each. Said before the fourth is typed rather than after it is sent.</p>
 */
export function remoteIsFull(turnsAsked: number): boolean {
  return turnsAsked >= REMOTE_TURNS;
}
